README.md
OpenAM Policy Agent for NGINX
=============================

# Download

## 20141119 Release

 * [Windows](https://www.osstech.co.jp/download/hamano/nginx/nginx_agent_20141119.msi)
 * [RHEL/CentOS 7 x86_64](https://www.osstech.co.jp/download/hamano/nginx/nginx_agent_20141119.el7.x86_64.zip)
 * [RHEL/CentOS 6 x86_64](https://www.osstech.co.jp/download/hamano/nginx/nginx_agent_20141119.el6.x86_64.zip)
 * [RHEL/CentOS 5 x86_64](https://www.osstech.co.jp/download/hamano/nginx/nginx_agent_20141119.el5.x86_64.zip)
 * [Debian/Ubuntu x86_64](https://www.osstech.co.jp/download/hamano/nginx/nginx_agent_20141119.deb.x86_64.zip)

# Requirement library

You need to install folowing library due to the nginx agent linked the
library dynamical.

for RHEL/CentOS:
~~~
# yum install nspr nss libxml2 pcpre openssl
~~~

for Debian/Ubuntu:
~~~
# apt-get install libnspr4 libnss3 libxml2 libpcre3 libssl1.0.0
~~~

# Agent Installation

## Steps

 1. Extract nginx_agent_YYYYMMDD.zip installation bits.

 ~~~
 # unzip nginx_agent_YYYYMMDD.zip -d /opt
 ~~~

 2. Goto nginx_agent

 ~~~
 # cd /opt/nginx_agent/
 ~~~

 3. Execute agentadmin.sh

 ~~~
 # ./bin/agentadmin.sh
 ~~~

 4. Follow the installation interactions and provide these details:

  - OpenSSO server URL
  - Agent Profile name
  - Agent Password

 5. Execute nginx
 ~~~
 # ./bin/nginx
 ~~~

# Agent Uninstallation

stop nginx and delete nginx_agent/ directory.

# Nginx Configuration

Specify OpenSSOAgentBootstrap.properties that agentadmin.sh has generated.
This is centralized agent configuration mode.

    http {
        ...
        am_boot_file "/path/to/OpenSSOAgentBootstrap.properties";

You can specify if you using local configuration mode.

        am_conf_file "/path/to/OpenSSOAgentConfiguration.properties";

If you want to use nginx as a reverce proxy.

    location / {
        ...
        proxy_pass        http://example.com:80/;
        proxy_set_header  X-Real-IP  $remote_addr;

Currently I recommend one worker process because multi process mode
does not work notification from OpenAM.

Pretty soon, I'll try solving the multi process problem. However Nginx
work very well with only one process.

    worker_processes  1;

If you found some problem, then please send me the debug logfile.

    error_log logs/error.log debug_http;


# Build Instructions

1. Install Dependencies

 for RHEL/CentOS:
 ~~~
 # yum install zlib-devel nspr-devel nss-devel libxml2-devel pcre-devel openssl-dev
 ~~~

 for Debian/Ubuntu:
 ~~~
 # apt-get install zlib1g-dev libnspr4-dev libnss3-dev libxml2-dev libpcre3-dev libssl-dev
 ~~~

2. Setup extlib

 You can download Agent SDK here:
 https://forgerock.org/downloads/openam-builds/

 ~~~
 $ mkdir extlib
 $ unzip -d extlib/common common_3_0_Linux_64bit.zip
 $ ln -s libamsdk.so.3 extlib/common/lib/libamsdk.so
 ~~~

3. Build

 ~~~
 $ make dist
 ~~~


CHANGELOG.md
### 20141119
 - Update to NGINX 1.7.7
 - Released Windows binary.

### 20141113
 - Update to NGINX 1.6.2
 - Dynamic linking with amsdk
 - Dynamic linking with OpenSSL
 - Released RHEL 5/6/7 binary.



conf/agentadmin.sh
#!/bin/sh

# Install script for nginx policy agent
#
# This script is just generating property files.
#
# Copyright (C) 2012 Tsukasa Hamano <hamano@osstech.co.jp>

#set -x

AGENT_ADMIN=`dirname "$0"`
AGENT_HOME=`readlink -f ${AGENT_ADMIN}/..`
CRYPT_UTIL="${AGENT_HOME}/bin/crypt_util"

AGENT_CONF_DIR=${AGENT_HOME}/conf
AGENT_BOOTSTRAP=${AGENT_CONF_DIR}/OpenSSOAgentBootstrap.properties
AGENT_CONFIG=${AGENT_CONF_DIR}/OpenSSOAgentConfiguration.properties
AGENT_BOOTSTRAP_TEMPL=${AGENT_CONF_DIR}/OpenSSOAgentBootstrap.template
AGENT_CONFIG_TEMPL=${AGENT_CONF_DIR}/OpenSSOAgentConfiguration.template

#RANDOM_SOURCE=/dev/random
# for debugging
RANDOM_SOURCE=/dev/urandom

usage(){
    cat << EOF

usage: $0 <option> [<arguments>]

The available options are:
--install: Installs a new Agent instance.

EOF
}

generate_key(){
    # EL6 have /usr/bin/base64 in coreutils
    # But the script need to working on other platform.
    head -c 24 ${RANDOM_SOURCE} | base64
}

agent_install_input(){
    cat << EOF
************************************************************************
Welcome to the OpenSSO Policy Agent for NGINX
************************************************************************

EOF

    echo 'Enter the URL where the OpenAM server is running.'
    echo 'Please include the deployment URI also as shown below:'
    echo '(http://opensso.sample.com:58080/opensso)'
    while [ -z ${OPENAM_URL} ]; do
        read -p "OpenSSO server URL: " OPENAM_URL
    done

    echo 'Enter the Agent profile name'
    while [ -z ${AGENT_PROFILE_NAME} ]; do
        read -p "Agent Profile Name: " AGENT_PROFILE_NAME
    done

    echo 'Enter the password to be used for identifying the Agent.'
    echo '*THIS IS NOT PASSWORD FILE*'
    stty -echo
    while [ -z ${AGENT_PASSWORD} ]; do
        read -p "Agent Password: " AGENT_PASSWORD
        echo
    done
    stty echo

    cat << EOF
-----------------------------------------------
SUMMARY OF YOUR RESPONSES
-----------------------------------------------
OpenSSO server URL : ${OPENAM_URL}
Agent Profile name : ${AGENT_PROFILE_NAME}
EOF
    echo 'Continue with Installation?'
    while [ -z ${CONFIRM} ]; do
        read -p "[y/N]: " CONFIRM
    done
    if [ ${CONFIRM} != "y" ]; then
        exit
    fi
}

agent_install(){
    AGENT_ENCRYPT_KEY=`generate_key`
    AGENT_ENCRYPTED_PASSWORD=`${CRYPT_UTIL} ${AGENT_PASSWORD} ${AGENT_ENCRYPT_KEY}`
    AGENT_LOGS_DIR="${AGENT_HOME}/logs/"
    sed -e "s%@OPENAM_URL@%${OPENAM_URL}%" \
        -e "s%@AGENT_PROFILE_NAME@%${AGENT_PROFILE_NAME}%" \
        -e "s%@AGENT_ENCRYPTED_PASSWORD@%${AGENT_ENCRYPTED_PASSWORD}%" \
        -e "s%@AGENT_ENCRYPT_KEY@%${AGENT_ENCRYPT_KEY}%" \
        -e "s%@AGENT_LOGS_DIR@%${AGENT_LOGS_DIR}%" \
        ${AGENT_BOOTSTRAP_TEMPL} > ${AGENT_BOOTSTRAP}

}

agent_install_input
agent_install



windows/agentadmin.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <Windows.h>
#include <wincrypt.h>

#include <iostream>
#include <string>
#include <fstream>
using namespace std;

#pragma comment(lib, "advapi32.lib")
#pragma comment(lib, "crypt32.lib")
#pragma comment(lib, "shell32.lib")

string generateKey(){
    HCRYPTPROV hProv = 0;
    BYTE agentEncryptKey[24];
    char key[48];
    DWORD keyLen;
    if(!CryptAcquireContext(&hProv, NULL, NULL, PROV_RSA_FULL,
                            CRYPT_VERIFYCONTEXT | CRYPT_SILENT)){
        perror("error at CryptAcquireContext()");
        exit(EXIT_FAILURE);
    }

    if(!CryptGenRandom(hProv, sizeof(agentEncryptKey), agentEncryptKey)){
        perror("error at CryptGenRandom()");
        exit(EXIT_FAILURE);
    }
    if(!CryptReleaseContext(hProv, 0)){
        perror("error at CryptReleaseContext()");
        exit(EXIT_FAILURE);
    }
    if(!CryptBinaryToString(agentEncryptKey, sizeof(agentEncryptKey),
                            CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF,
                            key, &keyLen)){
        perror("error at CryptBinaryToString()");
        exit(EXIT_FAILURE);
    }
    return string(key, keyLen);
}

string cryptPassword(string cryptCmd, string password, string key){
    string cmd = "\"" + cryptCmd + "\" " + password + " " + key;
    //cout << "cmd: " << cmd << "\n";
    FILE *pipe = _popen(cmd.c_str(), "r");
    char buf[1024];
    if(!pipe){
        perror("error at popen()");
        exit(EXIT_FAILURE);
    }
    fgets(buf, sizeof(buf), pipe);
    return string(buf);
}

int main(int argc, char *argv[]){
    char path[_MAX_PATH];
    GetModuleFileName(NULL, path, sizeof(path));
    string modulePath = string(path);
    string baseDir = modulePath.substr(0, modulePath.find_last_of("\\/"));
    string logsDir = baseDir + "\\logs";
    string confDir = baseDir + "\\conf";
    string cryptCmd = baseDir + "\\bin\\cryptit.exe";
    string bootstrapFile = confDir + "\\OpenSSOAgentBootstrap.properties";
    string bootstrapTmpl = confDir + "\\OpenSSOAgentBootstrap.template";
    string openamUrl;
    string agentProfileName;
    string agentPassword;
    HANDLE hStdin = GetStdHandle(STD_INPUT_HANDLE);
    DWORD mode = 0;
    GetConsoleMode(hStdin, &mode);

    cout << "\
************************************************************************\n\
Welcome to the OpenSSO Policy Agent for NGINX\n\
************************************************************************\n\
";
    cout << "Enter the URL where the OpenAM server is running.\n";
    cout << "Please include the deployment URI also as shown below:\n";
    cout << "(http://openam.sample.com:58080/openam)\n";
    do{
        cout << "OpenAM server URL: ";
        getline(cin, openamUrl);
    }while(openamUrl.empty());

    cout << "Enter the Agent profile name\n";
    do{
        cout << "Agent Profile Name: ";
        getline(cin, agentProfileName);
    }while(agentProfileName.empty());

    cout << "Enter the password to be used for identifying the Agent.\n";
    cout << "*THIS IS NOT PASSWORD FILE*\n";
    SetConsoleMode(hStdin, mode & (~ENABLE_ECHO_INPUT));
    do{
        cout << "Agent Password: ";
        getline(cin, agentPassword);
        cout << "\n";
    }while(agentPassword.empty());
    SetConsoleMode(hStdin, mode);

    string agentEncryptKey = generateKey();
    string agentEncryptedPassword = cryptPassword(cryptCmd,
                                                  agentPassword,
                                                  agentEncryptKey);

    cout << "\
-----------------------------------------------\n\
SUMMARY OF YOUR RESPONSES\n\
-----------------------------------------------\n\
";
    cout << "OpenSSO server URL : " << openamUrl << endl;
    cout << "Agent Profile name : " << agentProfileName << endl;
    cout << endl;
    cout << "Continue with Installation?\n";
    string input;
    do{
        cout << "[y/N]: ";
        getline(cin, input);
        if(!input.compare("N") || !input.compare("n")){
            exit(EXIT_SUCCESS);
        }
        if(!input.compare("Y") || !input.compare("y")){
            break;
        }
    }while(1);


    ifstream ifs(bootstrapTmpl);
    if(!ifs){
        perror("Cannot open template file");
        exit(EXIT_FAILURE);
    }
    ofstream ofs(bootstrapFile);
    if(!ofs){
        perror("Cannot open output file");
        exit(EXIT_FAILURE);
    }
    string line;
    int index;
    while(getline(ifs, line)){
        index = line.find("@OPENAM_URL@");
        if(index > 0){
            line.replace(index, 12, openamUrl);
        }
        index = line.find("@AGENT_PROFILE_NAME@");
        if(index > 0){
            line.replace(index, 20, agentProfileName);
        }
        index = line.find("@AGENT_ENCRYPTED_PASSWORD@");
        if(index > 0){
            line.replace(index, 26, agentEncryptedPassword);
        }
        index = line.find("@AGENT_ENCRYPT_KEY@");
        if(index > 0){
            line.replace(index, 19, agentEncryptKey);
        }
        index = line.find("@AGENT_LOGS_DIR@");
        if(index > 0){
            line.replace(index, 16, logsDir);
        }
        ofs << line << endl;
    }
    ifs.close();
    ofs.close();
    cout << "Generated " << bootstrapFile << endl;
    cout << "Enter any key to exit." << endl;
    // pause
    getline(cin, input);
    return EXIT_SUCCESS;
}
